### PR TITLE
Add support for custom weight decay optimizer

### PR DESCRIPTION
**Pull request by spetravic**
Creates a modified AdamW (that inherits from the standard AdamW) as suggested in _[Efficient deep reinforcement learning requires regulating overfitting](https://arxiv.org/pdf/2304.10466.pdf)_.

ModifiedAdamW disables weight decay for the bias, additional modules (such as LayerNorm), and specific named layers (e.g. the paper suggests excluding the final layer in the Q networks, which can be done by passing `layers_to_exclude = ["final_layer"]`). 

The function for separating the modules was adapted from a [PyTorch minGPT implementation](https://github.com/karpathy/minGPT/blob/3ed14b2cec0dfdad3f4b2831f2b4a86d11aef150/mingpt/model.py#L136) that is MIT licensed.

### CODE REVIEW

File: emote/optimizers.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+from typing import List, Optional, Set, Tuple
+
+import torch
+
+
+def separate_modules_for_weight_decay(
+    network: torch.nn.Module,
+    whitelist_weight_modules: Tuple[torch.nn.Module],
+    blacklist_weight_modules: Tuple[torch.nn.Module],
+    layers_to_exclude: Optional[List[str]] = None,
```
Comment (tgolsson) [inline emote/optimizers.py:14]: If you add `from __future__ import annotations` you can write this as:
```suggestion
    whitelist_weight_modules: tuple[torch.nn.Module, ...],
    blacklist_weight_modules: tuple[torch.nn.Module, ...],
    layers_to_exclude: set[str] | None = None,
```

which is a bit more modern!

File: emote/optimizers.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    Returns:
+        Tuple(Set, Set): Sets of modules with and without weight decay.
+    """
+
+    decay = set()
+    no_decay = set()
+
+    for mn, m in network.named_modules():
+        for pn, p in m.named_parameters():
+            fpn = "%s.%s" % (mn, pn) if mn else pn
```
Comment (tgolsson) [inline emote/optimizers.py:33]: Do you have to pay per letter? :D

File: emote/optimizers.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            if layers_to_exclude is not None and mn in layers_to_exclude:
+                # Weights of excluded layers will NOT be weight decayed
+                no_decay.add(fpn)
+            elif pn.endswith("bias"):
+                # Biases will not be decayed
+                no_decay.add(fpn)
+            elif pn.endswith("weight") and isinstance(m, blacklist_weight_modules):
+                # Weights of blacklist modules will NOT be weight decayed
+                no_decay.add(fpn)
+            elif pn.endswith("weight") and isinstance(m, whitelist_weight_modules):
```
Comment (tgolsson) [inline emote/optimizers.py:44]: I don't understand the usage of `isinstance` here! Either the type annotation lies or the code is wrong. You're taking a list of objects, but you're checking if m is a type. Should `whitelist_weight_modules` be `tuple[Type[nn.Module]]`?

File: emote/optimizers.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+        param_dict = {pn: p for pn, p in network.named_parameters()}
+
+        optim_groups = [
+            {
+                "params": [param_dict[pn] for pn in sorted(list(decay))],
+                "weight_decay": weight_decay,
+            },
+            {
+                "params": [param_dict[pn] for pn in sorted(list(no_decay))],
```
Comment (tgolsson) [inline emote/optimizers.py:90]: Is the order important here? It seems to me it wouldn't matter... :-) 

Also, if we need them to be lists here, why not make them lists and sort in the helper function?

File: tests/test_modified_adamw.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        return self.final_layer(self.encoder(x))
+
+
+def test_module_separation():
+    """Validate that all parameters are added to either "decay" or "no_decay"."""
+    q = QNet(2, 2, [2, 2])
+
+    decay, no_decay = separate_modules_for_weight_decay(
+        q,
+        whitelist_weight_modules=(torch.nn.Linear),
```
Comment (tgolsson) [inline tests/test_modified_adamw.py:39]: probably missing commas here as well

File: tests/test_modified_adamw.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        else:
+            no_wd += 1
+
+    assert (
+        wd == 1
+    ), f"There should be one group that has weight decay, but there are {wd}"
+
+    assert (
+        no_wd == 1
+    ), f"There should be one group that has weight decay, but there are {no_wd}"
```
Comment (tgolsson) [inline tests/test_modified_adamw.py:83]: Very nice tests! Could you reverse some of these tests and also pass in *bad* data and ensure it fails? There's the pattern

```python
a = {}
with pytest.raises(KeyError):
     a['foo'] # exception gets caught and test passes, but fails the test if the exception isn't caught 
```
You can also ensure the message is what you want using regex, like this:

```python
with pytest.raises(
    ValueError,
    match=r"My custom message: '[a-z]+'"
):
    raise ValueError("My custom message: 'foobar'")    
```

**Review by tgolsson — CHANGES_REQUESTED** [review]
Looks very nice! Super clean, with some nitpicks here and there!

File: emote/optimizers.py (hunk 2 above)
Comment (tgolsson) [reply emote/optimizers.py:33]: Also use f-strings!

**Review by singhblom — APPROVED** [review]
@tgolsson gave good feedback which I agree with, but apart from that it looks super nice. Looking forward to seeing the results of this. 👍🏼

File: emote/optimizers.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        layers_to_exclude: List[str] = None,
+    ):
+        decay, no_decay = separate_modules_for_weight_decay(
+            network,
+            whitelist_weight_modules,
+            blacklist_weight_modules,
+            layers_to_exclude,
+        )
+
+        param_dict = {pn: p for pn, p in network.named_parameters()}
```
[the remark this answers is not in the document]
Comment (spetravic) [reply emote/optimizers.py:82]: `network.named_parameters()` returns a generator, so the optimizer groups would have to be created in a different way. Converting to a dict and using the way it's written now seems cleaner to me than iterating through the modules with `next()` , but I could change it to use the generator instead if that is the preference.

[the remark this answers is not in the document]
Comment (spetravic) [reply emote/optimizers.py:82]: Yes, that works :)
